### PR TITLE
chore: handle exit code when there is a diff

### PR DESCRIPTION
**Pull request by Lou1415926**
https://github.com/aws/copilot-cli/pull/4662#pullrequestreview-1353462792 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

**Review by KollaAdithya — APPROVED** [review]

File: internal/pkg/cli/env_package.go [hunk 1]
```diff
@@ -192,7 +193,15 @@ func (o *packageEnvOpts) Execute() error {
 		return fmt.Errorf("generate CloudFormation template from environment %q manifest: %v", o.envName, err)
 	}
 	if o.showDiff {
-		return diff(packager, res.Template, o.diffWriter)
+		if err := diff(packager, res.Template, o.diffWriter); err != nil {
+			var errNonEmptyDiff errNonEmptyDiff
+			if errors.Is(err, &errNonEmptyDiff) {
```
Comment (dannyrandall) [inline internal/pkg/cli/env_package.go]: i think this should be `errors.As`

File: internal/pkg/cli/svc_deploy.go [hunk 2]
```diff
@@ -582,16 +586,35 @@ func (e *errFeatureIncompatibleWithEnvironment) RecommendActions() string {
 
 }
 
+type errNonEmptyDiff struct{}
+
+func (e *errNonEmptyDiff) Error() string {
+	return "diff detected "
```
Comment (dannyrandall) [inline internal/pkg/cli/svc_deploy.go]:
```suggestion
	return "diff detected"
```

File: internal/pkg/cli/svc_deploy.go [hunk 3]
```diff
@@ -582,16 +586,35 @@ func (e *errFeatureIncompatibleWithEnvironment) RecommendActions() string {
...
+func (e *errNonEmptyDiff) ExitCode() int {
+	return 1
+}
+
 func diff(differ templateDiffer, tmpl string, writer io.Writer) error {
 	out, err := differ.DeployDiff(tmpl)
 	if err != nil {
 		return err
 	}
+	var nonDiff bool
```
Comment (dannyrandall) [inline internal/pkg/cli/svc_deploy.go]:
```suggestion
	var noDiff bool
```

File: internal/pkg/cli/svc_deploy.go [hunk 4]
```diff
@@ -582,16 +586,35 @@ func (e *errFeatureIncompatibleWithEnvironment) RecommendActions() string {
...
 	}
 	_, err = writer.Write([]byte(out))
-	return err
+	if err != nil {
+		return err
+	}
+	if nonDiff {
+		return nil
+	}
+	return &errNonEmptyDiff{}
```
Comment (dannyrandall) [inline internal/pkg/cli/svc_deploy.go]: non empty diff confuses me haha...could we change the wording/logic to `noDiff`?

Comment (Lou1415926) [reply internal/pkg/cli/svc_deploy.go]:
```
out, err := differ.DeployDiff(tmpl)
	if err != nil {
		return err
	}
	if out != "" {
		if _, err := writer.Write([]byte(out)); err != nil {
			return err
		}
		return &errHasDiff{}
	}
	if _, err := writer.Write([]byte("No changes.\n")); err != nil {
		return err
	}
	return nil
```

Better?

File: internal/pkg/cli/svc_deploy.go [hunk 5]
```diff
@@ -582,16 +586,32 @@ func (e *errFeatureIncompatibleWithEnvironment) RecommendActions() string {
...
 		return err
 	}
-	if out == "" {
-		out = "No changes.\n"
+	if out != "" {
+		if _, err := writer.Write([]byte(out)); err != nil {
+			return err
+		}
+		return &errHasDiff{}
+	}
```
Comment (paragbhingre) [inline internal/pkg/cli/svc_deploy.go]: How about ?

```
if out, err := differ.DeployDiff(tmpl); err != nil {
	return err
} else if out != "" {
	if _, err := writer.Write([]byte(out)); err != nil {
		return err
	}
	return &errHasDiff{}
}
```

**Review by paragbhingre — APPROVED** [review]
LGTM! DNM just in case any comment left to be addressed from Danny. Feel free to remove it.

**Pull request merged**